File: 프로그래머스/3/214288. 상담원 인원/상담원 인원.cpp
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int min_total_wait = 1e9;

int calculate_wait_time(int k , const vector<int>& mentors_per_type, const vector<vector<int>>& reqs){
    vector<priority_queue<int, vector<int>, greater<int>>> mentor_queues(k + 1);

    for(int i = 1;i <= k; i++){
        for(int j = 0; j < mentors_per_type[i - 1]; ++j){
            mentor_queues[i].push(0);
        }
    }

    int total_wait = 0;
    for(const auto& req : reqs){
        int start_time = req[0];
        int duration = req[1];
        int type = req[2];

        int earlist_finish_time = mentor_queues[type].top();
        mentor_queues[type].pop();

        if(earlist_finish_time > start_time){
            int wait_time = earlist_finish_time - start_time;
            total_wait += wait_time;

            mentor_queues[type].push(earlist_finish_time + duration);
        }else {
            mentor_queues[type].push(start_time + duration);
        }
    }
    return total_wait;
}

void find_combinations(int k, int remain_mentors, int start_idx, vector<int>& mentors_per_type, const vector<vector<int>>& reqs) {
    if(remain_mentors == 0){
        int current_wait = calculate_wait_time(k, mentors_per_type, reqs);
        min_total_wait = min(min_total_wait, current_wait);
        return;
    }

    for(int i = start_idx; i < k; i++){
        mentors_per_type[i]++;
        find_combinations(k, remain_mentors - 1, i, mentors_per_type, reqs);
        mentors_per_type[i]--;
    }
}

int solution(int k , int n , vector<vector<int>> reqs){
    vector<int> mentors_per_type(k, 1);

    find_combinations(k, n - k, 0, mentors_per_type, reqs);

    return min_total_wait;
}

File: 프로그래머스/3/214288. 상담원 인원/상담원 인원.js
function solution(k, n, reqs) {
  // 조합 + 큐(상담 유형 별)
  let answer = Infinity;

  const startConsult = (numberOfType) => {
    console.log(numberOfType)

    let candidate = 0; // 예비 정답

    const queue = {}; // 타입별 상담 종료 시간을 넣는 곳(내림차순 저장)
    for (let t = 1; t <= k; t++) {
      queue[t] = [];
    }

    for (const req of reqs) {
      if (candidate >= answer) break; // 현재 정답보다 크다면 그만해도 됨.

      const [startTime, consultTime, type] = req;

      // startTime 이전 상담 큐별로 정리시간!
      for (let t = 1; t <= k; t++) {
        let flag = false;
        while (!flag) {
          const size = queue[t].length;
          if (size > 0 && queue[t][size - 1] <= startTime) {
            // 마지막 요소가 startTime보다 작거나 같다면 제거!
            queue[t].pop();
          } else {
            flag = true;
          }
        }
      }

      // 해당 상담 처리하기
      const size = queue[type].length;

      let additionalTime = 0;
      if (size == numberOfType[type - 1]) {
        // 조금 기다려야 하는가?
        additionalTime = queue[type][size - 1] - startTime;
        queue[type].pop();
      }

      queue[type].push(startTime + consultTime + additionalTime); // 상담 진행
      queue[type].sort((a, b) => b - a); // 상담 들어간 배열 sort 진행

      candidate += additionalTime;
    }
    // console.log(candidate)

    answer = Math.min(answer, candidate); // 정답 비교 후 초기화
  };

  const combination = (numberOfType, remainMentor, startIndex) => {
    if (answer == 0) return;

    if (remainMentor == 0) {
      startConsult(numberOfType);
    } else {
      for (let i = startIndex; i < k; i++) {
        numberOfType[i]++;
        combination(numberOfType, remainMentor - 1, i);
        numberOfType[i]--;
      }
    }
  };

  combination(Array(k).fill(1), n - k, 0);

  return answer;
}
